src/Task.cc

```diff
@@ -1345,13 +1366,35 @@ void Task::flush_regs() {
   if (registers_dirty) {
     LOG(debug) << "Flushing registers for tid " << tid << " " << registers;
     auto ptrace_regs = registers.get_ptrace();
+#if defined(__i386__) || defined(__x86_64)
     if (ptrace_if_alive(PTRACE_SETREGS, nullptr, &ptrace_regs)) {
       /* It's ok for flush regs to fail, e.g. if the task got killed underneath
        * us - we just need to remember not to trust any value we would load
        * from ptrace otherwise */
       registers_dirty = false;
+      orig_syscallno_dirty = false;
+    }
+#elif defined(__aarch64__)
+    struct iovec vec = { &ptrace_regs,
+                          sizeof(ptrace_regs) };
+    if (ptrace_if_alive(PTRACE_SETREGSET, NT_PRSTATUS, &vec)) {
+      registers_dirty = false;
+    }
+#else
+    #error "Unknown archietcture"
+#endif
+  }
+#if defined(__aarch64__)
+  if (orig_syscallno_dirty) {
+    uintptr_t syscall = registers.original_syscallno();
+    struct iovec vec = { &syscall,
+                          sizeof(syscall) };
+    LOG(debug) << "Chaning syscall to " << syscall;
+    if (ptrace_if_alive(PTRACE_SETREGSET, NT_ARM_SYSTEM_CALL, &vec)) {
+      orig_syscallno_dirty = false;
     }
   }
+#endif
```

Comment:
```
#elif defined(__i386__) || defined(__x86_64__)
  ASSERT(t, !orig_syscallno_dirty);
#endif
```

Reply:
We can't really do that, right, because the ptrace above is allowed to fail?
At most we could do:
```
#elif defined(__i386__) || defined(__x86_64__)
  ASSERT(t, registers_dirty == orig_syscallno_dirty);
#endif
```

Reply:
Ah, yes, good point.

Reply:
Actually, that's not quite right either, because it is possible for the registers to be dirty, but orig_syscallno not to be dirty. I think the most we can do is maybe something like:
```
  ASSERT(this, !orig_syscallno_dirty || registers_dirty);
```

Reply:
maybe in the x86 `if (registers_dirty)` block add `else { ASSERT(this, !orig_syscallno_dirty); }` then?

Reply:
yep